### PR TITLE
Mzuevx/rnnt evaluator

### PR DESCRIPTION
**Pull request by mzuevx**
Custom evaluator for encoder-prediction-joint models

### CODE REVIEW

**Comment by eaidova** [comment]
Jenkins ok to test

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py [hunk 1]
```diff
@@ -0,0 +1,687 @@
+"""
+Copyright (c) 2018-2020 Intel Corporation
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py:2]:
```suggestion
Copyright (c) 2018-20201 Intel Corporation
```

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py [hunk 2]
```diff
@@ -0,0 +1,687 @@
...
+
+        if 'progress_reporter' in kwargs:
+            _progress_reporter = kwargs['progress_reporter']
+            _progress_reporter.reset(self.dataset.size)
+        else:
+            _progress_reporter = None if not check_progress else self._create_progress_reporter(
+                check_progress, self.dataset.size
+            )
+        for batch_id, (batch_input_ids, batch_annotation, batch_inputs, batch_identifiers) in enumerate(self.dataset):
+            rnnt_ref = np.load("./rnnt_ref.npz")
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py:86]: What is this??? Why path is constant?

**Comment by mzuevx** [comment]
@eaidova , DLSDK functionality implemented. ONNX_runtime part still untested and may be incorrectly implemented because ONNX models with dynamic axes can't be reshaped at runtime so I can't test it at all. I'll continue investigate this issue.

File: tools/accuracy_checker/accuracy_checker/adapters/audio_recognition.py [hunk 3]
```diff
@@ -681,6 +681,28 @@ def empty(lm=None, start_with_bos=True):
...
+        })
+        return parameters
+
+    def configure(self):
+        self.alphabet = self.get_value_from_config('alphabet') or ' ' + string.ascii_lowercase + '\''
+        self.alphabet = self.alphabet.encode('ascii').decode('utf-8')
+
+    def process(self, raw, identifiers=None, frame_meta=None):
+        decoded = ''.join(self.alphabet[t] for t in raw[0])
+        return [CharacterRecognitionPrediction(identifiers[0], decoded.upper())]
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/adapters/audio_recognition.py:705]: I think moving to uppercase or not should be adapter parameter.
Also need to check that you have only 1 identifier (gove assert if not), if model does not support batching

File: tools/accuracy_checker/accuracy_checker/data_readers/data_reader.py [hunk 4]
```diff
@@ -612,7 +612,6 @@ def read(self, data_id):
     def read_item(self, data_id):
         return DataRepresentation(*self.read_dispatcher(data_id), identifier=data_id)
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/data_readers/data_reader.py:614]: please return empty line (recommended for making code more readable, usually for classes used 2 line spacing)

File: tools/accuracy_checker/accuracy_checker/adapters/audio_recognition.py (hunk 3 above)
Comment (eaidova) [reply tools/accuracy_checker/accuracy_checker/adapters/audio_recognition.py:705]: Please add new adapter to docs

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py [hunk 5]
```diff
@@ -0,0 +1,669 @@
...
+from ..quantization_model_evaluator import create_dataset_attributes
+from ...adapters import create_adapter
+from ...config import ConfigError
+from ...launcher import create_launcher
+from ...utils import contains_all, contains_any, extract_image_representations, read_pickle, get_path
+from ...progress_reporters import ProgressReporter
+from ...logging import print_info
+
+
+class AutomaticSpeechRecognitionEvaluator(BaseEvaluator):
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py:33]: I do not see the difference in evaluator code between this and encoder-decoder (all diff incapsulated in the model), maybe you can just inherit it from that class for avoiding a lot of code duplication?

File: tools/accuracy_checker/accuracy_checker/preprocessor/audio_preprocessing.py [hunk 6]
```diff
@@ -15,6 +15,7 @@
 """
 
 import numpy as np
+import scipy.signal as dsp
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/preprocessor/audio_preprocessing.py:18]: please wrap to unsupported package, scipy is trouble maker for some platforms, we avoid to have it as explicit dependency if user do not want to use models where it is really required

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py (hunk 5 above)
Comment (eaidova) [reply tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py:33]: please also provide docs for your evaluator

File: tools/accuracy_checker/accuracy_checker/preprocessor/audio_preprocessing.py (hunk 6 above)
Comment (mzuevx) [reply tools/accuracy_checker/accuracy_checker/preprocessor/audio_preprocessing.py]: Fixed

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py (hunk 5 above)
Comment (mzuevx) [reply tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py]: Fixed

File: tools/accuracy_checker/accuracy_checker/data_readers/data_reader.py (hunk 4 above)
Comment (mzuevx) [reply tools/accuracy_checker/accuracy_checker/data_readers/data_reader.py]: Fixed

File: tools/accuracy_checker/accuracy_checker/adapters/audio_recognition.py (hunk 3 above)
Comment (mzuevx) [reply tools/accuracy_checker/accuracy_checker/adapters/audio_recognition.py:710]: Fixed

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+            if not contains_any(prediction, ['model', 'onnx_model']) and models_args:
+                prediction['model'] = models_args[1 if len(models_args) > 1 else 0]
+                prediction['_model_is_blob'] = is_blob
+            if not contains_any(joint, ['model', 'onnx_model']) and models_args:
+                joint['model'] = models_args[2 if len(models_args) > 2 else 0]
+                joint['_model_is_blob'] = is_blob
+            network_info.update({'encoder': encoder, 'prediction': prediction, 'joint': joint})
+        if not contains_all(network_info, ['encoder', 'prediction', 'joint']) and not delayed_model_loading:
+            raise ConfigError('network_info should contain encoder, prediction and joint fields')
+        # self.num_processing_frames = network_info['decoder'].get('num_processing_frames', 16)
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py:214]: please remove commented code

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py [hunk 8]
```diff
@@ -0,0 +1,456 @@
...
+        logits, logits_blob = self.joint.predict(None, inputs)
+        logits = logits_blob[:, 0, 0, :]
+        if not log_normalize:
+            return logits
+
+        # probs = F.log_softmax(logits, dim=len(logits.shape) - 1)
+        # return probs
+
+    def _get_last_symb(self, labels) -> int:
+        return self._SOS if len(labels) == 0 else labels[-1]
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py:341]: can this variable be named without uppercase?

**Comment by eaidova** [comment]
@mzuevx please fulfil pylint compliance and I'll approve and merge your changes

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py (hunk 8 above)
Comment (mzuevx) [reply tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py]: fixed

File: tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py (hunk 7 above)
Comment (mzuevx) [reply tools/accuracy_checker/accuracy_checker/evaluators/custom_evaluators/asr_encoder_prediction_joint_evaluator.py]: fixed

**Review by eaidova — APPROVED** [review]

**Pull request merged**